base/linalg/special.jl

```diff
@@ -141,6 +141,24 @@ for op in (:+, :-)
             end
         end
     end
+
+    for matrixtype1 in (:Diagonal,)
+        for (matrixtype2,matrixtype3) in ((:UpperTriangular,:UpperTriangular),
+                                          (:UnitUpperTriangular,:UpperTriangular),
+                                          (:LowerTriangular,:LowerTriangular),
+                                          (:UnitLowerTriangular,:LowerTriangular))
+            @eval begin
+                ($op)(A::($matrixtype1), B::($matrixtype2)) = ($op)(convert(($matrixtype3), A), B)
```

Comment:
This solution creates an unnecessary temporary array. This is in line with most of the definitions in `special.jl` but at some point it would be good to get rid of it. Something like
```jl
function (-)(A::Diagonal, B::StridedMatrix)
    C = -B
    for i = 1:size(C,1)
        C[i,i] += A.diag[i]
    end
    return C
end
```
in `diagonal.jl` and the
```jl
($op)(A::($matrixtype1), B::($matrixtype2)) = ($matrixtype3)(($op)(A,B.data))
```
in `special.jl` should accomplish that.